strawberryfields/apps/train/cost.py

```diff
@@ -0,0 +1,183 @@
+
+        >>> sample = [1, 1, 0, 0]
+        >>> cost.h_reparametrized(sample, params)
+        -1.6688383062813434
+
+        Args:
+            sample (array): the sample
+            params (array): the trainable parameters :math:`\theta`
+
+        Returns:
+            float: the cost function with respect to a given sample and set of trainable parameters
+        """
+        h = self.h(sample)
+        A = self.vgbs.A(params)
+        w = self.vgbs.embedding(params)
+        Id = np.eye(2 * self.vgbs.n_modes)
+
+        dets_numerator = np.linalg.det(Id - _Omat(A))
+        dets_denominator = np.linalg.det(Id - _Omat(self.vgbs.A_init))
+        dets = np.sqrt(dets_numerator / dets_denominator)
+
+        prod = np.prod(np.power(w, sample))
+
+        return h * dets * prod
+
+    def sample_difference_from_mean(self, sample: np.ndarray, params: np.ndarray) -> np.ndarray:
+        """Calculates the difference between an input sample and the vector of mean clicks or
+        photons by mode.
+
+        **Example usage:**
+
+        >>> sample = [1, 0, 0, 0]
+        >>> cost.sample_difference_from_mean(sample, params)
+        array([ 0.47187426,  0.4798068 ,  0.46717688, -0.53437824])
+
+        Args:
+            sample (array): the sample
+            params (array): the trainable parameters :math:`\theta`
+
+        Returns:
+            array: the difference
+        """
+        if self.vgbs.threshold:
+            n_diff = sample - self.vgbs.mean_clicks_by_mode(params)
+        else:
+            n_diff = sample - self.vgbs.mean_photons_by_mode(params)
+
+        return n_diff
+
+    def _gradient_one_sample(self, sample: np.ndarray, params: np.ndarray) -> np.ndarray:
+        """Evaluates the gradient equation on a single sample.
+
+        Args:
+            sample (array): the sample
+            params (array): the trainable parameters :math:`\theta`
+
+        Returns:
+            array: the one shot gradient
+        """
+        w = self.vgbs.embedding(params)
+        jac = self.vgbs.embedding.jacobian(params)
+
+        h = self.h_reparametrized(sample, params)
+        diff = self.sample_difference_from_mean(sample, params)
+        return h * (diff / w) @ jac
+
+    def gradient(self, params: np.ndarray, n_samples: int = 100) -> np.ndarray:
+        """Evaluates the gradient of the cost function.
+
+        The gradient is evaluated by finding an average of a function over a number ``n_samples`` of
+        samples generated from the VGBS system using the trainable parameters :math:`\theta`.
+
+        The averaged function is specified in `this paper <https://arxiv.org/abs/2004.04770>`__.
+
+        **Example usage:**
+
+        >>> cost.gradient(params, 100)
+        array([ 0.10880756, -0.1247146 ,  0.12426481, -0.13783342])
+
+        Args:
+            params (array): the trainable parameters :math:`\theta`
+            n_samples (int): the number of GBS samples used to average the gradient
```

Comment:
```suggestion
            n_samples (int): the number of GBS samples used in the gradient estimation
```
Is it correct that we re averaging the gradient? I think it's more accurate that the gradient is an average, which we estimate by sampling